core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -2375,25 +2375,37 @@ private void checkStorageTableNotRedirected(QualifiedObjectName source)
 
         private void analyzeFiltersAndMasks(Table table, QualifiedObjectName name, RelationType relationType, Scope accessControlScope)
         {
-            ImmutableList.Builder<ColumnSchema> columnSchemaBuilder = ImmutableList.builder();
+            ImmutableMap.Builder<String, ColumnSchema> columnSchemaBuilder = ImmutableMap.builder();
             for (int index = 0; index < relationType.getAllFieldCount(); index++) {
                 Field field = relationType.getFieldByIndex(index);
-                field.getName().ifPresent(fieldName -> columnSchemaBuilder.add(ColumnSchema.builder()
+                field.getName().ifPresent(fieldName -> columnSchemaBuilder.put(fieldName, ColumnSchema.builder()
                         .setName(fieldName)
                         .setType(field.getType())
                         .setHidden(field.isHidden())
                         .build()));
             }
-            List<ColumnSchema> columnSchemas = columnSchemaBuilder.build();
+            Map<String, ColumnSchema> columnSchemas = columnSchemaBuilder.buildOrThrow();
 
-            Map<ColumnSchema, ViewExpression> masks = accessControl.getColumnMasks(session.toSecurityContext(), name, columnSchemas);
+            Map<ColumnSchema, ViewExpression> masks = accessControl.getColumnMasks(session.toSecurityContext(), name, ImmutableList.copyOf(columnSchemas.values()));
 
-            for (ColumnSchema columnSchema : columnSchemas) {
-                Optional.ofNullable(masks.get(columnSchema)).ifPresent(mask -> {
-                    if (checkCanSelectFromColumn(name, columnSchema.getName())) {
```

Comment:
I don't know why we were doing a check here in a first place. See `io.trino.sql.analyzer.Analyzer#analyze(io.trino.sql.tree.Statement, io.trino.sql.analyzer.QueryType)`. We used to have similar problem in the past that same tables and columns were checked many times, so we collect then during analysis and check once at the end. I guess we need follow the approach.

Reply:
I'm not sure either, but I guess it's because we don't want to do analysis on things the user doesn't have access to because it may in fact be creating a risk of information leakage. I can't think of a scenario off the top of my head, but I think caution is warranted here.

Reply:
Do you think if we would have https://github.com/trinodb/trino/pull/15131 then we would be able to solve https://github.com/trinodb/trino/issues/4496? Also I believe with that approach we would be able to access control checks in `io.trino.sql.analyzer.Analyzer#analyze(io.trino.sql.tree.Statement, io.trino.sql.analyzer.QueryType)` as we could simply do then sooner. But what we need to have is a test for number of access control checks to make sure we don't regress it.